Feature/page object pattern

rodikno opened the pull request:
1) Added answers to the page object section task
2) Created initial structure for using PageObject in the Main Solution

rodikno committed:
Initial file and content for answers

rodikno committed:
3 Answers added

rodikno committed:
Changed indentation for 2 spaces

rodikno committed:
Prepaired initial structure for moving Main Solution to Page Object pattern usage

rodikno committed:
Moved page objects under features/support folder, added lost method

rodikno committed:
OMG! First scenario with usage of PAge Object works!

rodikno committed:
3 scenarious migrated to PageObject

rodikno committed:
In addition to the previous commit, 3 scenarious migrated to PO

rodikno committed:
Change password scenrio, used steps within step definitions

rodikno committed:
Create project scenario with some cool trick (putting dynamic part into page_url)

d-konovalov commented on Main_Solution/our_module.rb:
```diff
@@ -191,6 +187,14 @@ def is_issue_a_bug?(issue)
   end
 
+  def is_issue_watched?
+    if find_element_by_css('a.icon-fav').displayed?
```
we can leave only this line

rodikno replied:
Fixed (not using this anymore)

d-konovalov commented on features/step_definitions/redmine_steps.rb:
```diff
@@ -7,59 +7,68 @@
 
 Then(/^I am logged in$/) do
-  expect(find_element_by_class('user').text).to eql @user.login
+  active_username = on(MyPage).top_menu.active_user_element.attribute 'text'
```
active_user_element.text should work

rodikno replied:
Yep, much more simple, thanks)

d-konovalov commented on features/step_definitions/redmine_steps.rb:
```diff
@@ -7,59 +7,68 @@
+
+  success_message = on(MyAccountPage).success_message_element
+
+  expect(@current_page.page_url_value).to eql on(MyAccountPage).page_url_value
```
page_url_value will ask url value from instance of a class of the page but not from browser so you compare here the url string with itself actually )

rodikno replied:
Fixed

d-konovalov commented on features/step_definitions/redmine_steps.rb:
```diff
@@ -7,59 +7,68 @@
-  success_message = find_element_by_id('flash_notice')
-  expect(success_message).to be_displayed
+  success_message = on(MyAccountPage).success_message_element
+  expect(success_message).to be_visible
```
if you would define method like 'has_success_message?' in the MyAccountPage class
you could use this kind of expectation:

expect(on(MyAccountPage)).to have_success_message

it's not a mistake just small improvement

rodikno replied:
Now BasicPage has this method: `has_success_message?` and also all other Page Classes are descendants of BasicPage so always using it for success_message expectations now, thanks)

d-konovalov commented on features/step_definitions/redmine_steps.rb:
```diff
@@ -7,59 +7,68 @@
 And(/^I can login with a new password$/) do
-  log_out
-  log_in(@user.login, @user.password)
+  step 'I log out'
```
step is well know cucumber feature 
but there is one danger case possible:
what if you'll need to refactor some of your steps and i'll update them
RubyMine can't rename steps' titles with one action (yet) so it can bring you to go and find change you have to make until your tests work again 
you can try to use it if you promise to be careful )

rodikno replied:
Refactored step_definitions and removed `step` usages.

Btw I was looking for a workaround (how to reuse already defined steps in step definitions to keep them DRY) and found this: https://makandracards.com/makandra/18905-how-to-not-repeat-yourself-in-cucumber-scenarios

There's a `many_steps.rb` helper proposed by author so it supposed to solve problem:
> When calling steps with multiple lines of Cucumber you lose meaningful stack traces. If a step fails you will always get the same file and line number: The one where you call steps. This is a big deal in practice. To remedy this, check out our many_steps helper (see below).

What do you think about this practice, maybe there's any graceful way to avoid code duplication if i need to gather some steps within one to keep scenario simple?

d-konovalov commented on features/support/page_objects/top_menu_section.rb:
```diff
@@ -0,0 +1,9 @@
+  link(:log_out, :css => 'a.logout')
+  link(:register, :css => 'a.register')
+  link(:my_account, :css => 'a.my-account')
+  link(:active_user, :css => 'div#loggedas > a.user.active')
```
all is good but could you reduce css locator length as much as possible?

rodikno replied:
Changed to `#loggedas > a`, looks like the shortest as for me now :)

rodikno committed:
Open random project scenario finished

rodikno committed:
Create version scenario reworked

d-konovalov commented on features/step_definitions/redmine_steps.rb:
```diff
@@ -73,12 +78,22 @@
-  open_random_project(@project[:name], retries_count.to_i)
+  random_project_url = "http://demo.redmine.org/projects/#{@project[:name]}"
+
+  i = 0
```
Too much of technical details in step definition 
Pease think about possibility to hide it somewhere
(sometimes it's ok to define some helper methods then wrap them up in module and include it in env.rb for example)

rodikno replied:
Now separate method `open_url_with_retries` added for this but some logic is still here, in step definition so step definition looks like:

```ruby
When(/^I try to open random project with (\d+) retries$/) do |retries_count|
  random_name = Faker::Hipster.word.capitalize
  random_project_url = BasicPage::BASIC_URL + '/projects/' + random_name

  project_exists = open_url_with_retries(random_project_url, retries_count.to_i)

  if project_exists
    puts "Project with name [#{random_name}] already exists"
  else
    @project.name = random_name
    visit(CreateProjectPage).create_project(@project)
    puts "New project [#{@project.name}] created"
  end
end
```

I've thought about making it shorter, but see no obvious way how to hande this without affecting readability, what do you think about this?

rodikno committed:
Changed project structure: moved pages to ./pages, moved helpers to ./helpers

rodikno committed:
Clean up in hooks.rb: user require_all for pages and helper resources, added separate classes for test subjects (user, project, issue)

rodikno committed:
Refactored steps and step definitions for User

rodikno committed:
Refactored [Create project if not found] scenario (which is actually not a test scenario but a fallback logic implementation)

rodikno committed:
Create issue scenario refactored

rodikno committed:
Last scenario complete (Add user as a watcher to an issue)

rodikno committed:
Refactoring:
- Moved pages to corresponding folders depending on feature
- All Page Classes are now descendants of BasicPage
- Moved TopMenuSection to BasicPage since it's present on all pages of the site
- A blank line at the beginning of class definition to make it more readable
- Removed unused pages

rodikno committed:
Changed all page_url methods in Page Classes for BASIC_URL + [specific_part], where BASIC_URL is defined in BasicPage class

rodikno committed:
Changed selector to a shorter one for :active_user element

d-konovalov commented on Main_Solution/our_module.rb:
```diff
@@ -1,6 +1,5 @@
 require 'faraday'
 require 'C:\_projects\automation_course\Main_Tasks\Exceptions\my_exceptions.rb'
```
this line looks bad (
if I run your tests on my machine this won't work

rodikno replied:
Fixed, changed to relative path

d-konovalov commented on features/step_definitions/redmine_steps.rb:
```diff
@@ -42,110 +25,95 @@
-  username_in_watchers_list = find_element_by_css("li.user-#{@user.id}")
-
-  expect(username_in_watchers_list).to be_displayed
+  @browser.navigate.refresh
```
@current_page.refresh would work as well

rodikno replied:
Yep, works)

d-konovalov commented on pages/issue_pages/issue_details_page.rb:
```diff
@@ -0,0 +1,53 @@
+  end
+
+  def close_issue
+    self.edit_issue
```
you need 'self' usually when you call methods with '=' at the end since ruby will think that you're creating a variable
in this case just 'edit_issue' would be enough

rodikno replied:
Changed all method calls which (except used with `=`) to be called without `.self`

d-konovalov commented on pages/issue_pages/issue_details_page.rb:
```diff
@@ -0,0 +1,53 @@
+  end
+
+  def is_closed?
+    self.issue_status == 'Closed' ? true : false
```
'self' is redundant as well

rodikno replied:
Fixed

rodikno committed:
Changed absolute path to relative

rodikno committed:
Updated .gitignore

rodikno committed:
Changed @browser to @current_page for refreshing

rodikno committed:
Removed .self from element method where it's not needed
